### ISSUE TITLE
Loading marshalled Java enums with anonymous classes does not work

### ISSUE THREAD

**Issue by jmiettinen** [jruby/jruby#3526]
If we have a class
```java
public enum SimpleCase {
    FIZZ,
    BUZZ {
        @Override
        public boolean foo() {
            return true;
        }
    };
    public boolean foo() { return false; }
}
```
Then, running `CLASSPATH="." ruby -e 'Java::SimpleCase.values.each { |e| Marshal.load(Marshal.dump(e)) }'` fails for `BUZZ` with error
```
ArgumentError: undefined class/module Java::Default::SimpleCase::1
    load at org/jruby/RubyMarshal.java:148
  (root) at -e:1
    each at org/jruby/java/proxies/ArrayJavaProxy.java:154
  (root) at -e:1
```
I am not sure if this should work, but it did in `1.7.19` and no longer works in `1.7.20` and later ones.

**Comment by kares** [comment]
yes this should work just fine - we'll be looking at it, thanks for the report!

**Comment by kares** [comment]
here's the deal - anonymous Java classes used to look a bit weird on the Ruby end e.g. `Java::JavaUtilConcurrent::TimeUnit:: ` which was changed at https://github.com/jruby/jruby/commit/2c50057a9a833a4d97d0c390f737fdc9e4373058 so they look e.g. `Java::JavaUtilConcurrent::TimeUnit::1` since. marshalling failed somehow correctly for older 1.7.x  :
```
jruby-1.7.4 :015 > secs = Marshal.load Marshal.dump [ java.util.concurrent.TimeUnit::SECONDS ]
ArgumentError: Java::JavaUtilConcurrent:: does not refer class
	from org/jruby/RubyMarshal.java:148:in `load'
	from (irb):15:in `evaluate'
	from org/jruby/RubyKernel.java:1093:in `eval'
	from org/jruby/RubyKernel.java:1489:in `loop'
	from org/jruby/RubyKernel.java:1254:in `catch'
	from org/jruby/RubyKernel.java:1254:in `catch'
	from /opt/local/rvm/rubies/jruby-1.7.4/bin/irb:13:in `(root)'
```
yet for some reason marshalling works these works on 1.7.19 but does not serialize the "correct" class :
```
jruby-1.7.19 :008 > java.util.concurrent.TimeUnit::SECONDS.class
 => Java::JavaUtilConcurrent::TimeUnit:: 
jruby-1.7.19 :005 > secs = Marshal.load Marshal.dump java.util.concurrent.TimeUnit::SECONDS
 => #<Java::JavaUtilConcurrent::TimeUnit:0x704c6d5c> 
jruby-1.7.19 :006 > secs.class
 => Java::JavaUtilConcurrent::TimeUnit 
jruby-1.7.19 :007 > secs.class == java.util.concurrent.TimeUnit::SECONDS.class
 => false 
```
this would be easily fixed if the `RubyClass` meta-class was not abused with Java integration and we had a `JavaClass` sub-class used for Java classes (a big rewrite I am planning for a while but could not get to :) 

... really hesitant to revert this as the previous behaviour is not correct and might lead to ugly bugs ;(

**Comment by jmiettinen** [comment]
Ah yeah, the earlier version did not work correctly either. I never noticed that because we were just accidentally marshalling and unmarshalling enums and that started to fail.

So the wrong result was never used.

**Comment by kares** [comment]
working fine on 9K - calling it a wont-fix for jruby 1.7.x as its better to fail than de-serialize improperly.
